utils/PointerTracker.ts

```diff
@@ -0,0 +1,234 @@
+  constructor (private _element: HTMLElement, {
+    start = () => true,
+    move = noop,
+    end = noop
+  }: PointerTrackerCallbacks) {
+    this._startCallback = start;
+    this._moveCallback = move;
+    this._endCallback = end;
+
+    // Bind listener methods
+    this._pointerStart = this._pointerStart.bind(this);
+    this._touchStart = this._touchStart.bind(this);
+    this._move = this._move.bind(this);
+    this._pointerEnd = this._pointerEnd.bind(this);
+    this._touchEnd = this._touchEnd.bind(this);
+
+    // Add listeners
+    if (self.PointerEvent) {
+      this._element.addEventListener('pointerdown', this._pointerStart);
+    } else {
+      this._element.addEventListener('mousedown', this._pointerStart);
+      this._element.addEventListener('touchstart', this._touchStart);
+      this._element.addEventListener('touchmove', this._move);
+      this._element.addEventListener('touchend', this._touchEnd);
+    }
+  }
+
+  /**
+   * Call the start callback for this pointer, and track it if the user wants.
+   *
+   * @param pointer Pointer
+   * @param event Related event
+   * @returns Whether the pointer is being tracked.
+   */
+  private _triggerPointerStart (pointer: Pointer, event: PointerEvent | MouseEvent | TouchEvent): boolean {
+    if (!this._startCallback(pointer, event)) return false;
+    this.currentPointers.push(pointer);
+    this.startPointers.push(pointer);
+    return true;
+  }
+
+  /**
+   * Listener for mouse/pointer starts. Bound to the class in the constructor.
+   *
+   * @param event This will only be a MouseEvent if the browser doesn't support
+   * pointer events.
+   */
+  private _pointerStart (event: PointerEvent | MouseEvent) {
+    // Only interested in left-button presses.
+    if (event.button !== 0) return;
+    if (!this._triggerPointerStart(new Pointer(event), event)) return;
+
+    // Add listeners for additional events.
+    // The listeners may already exist, but no harm in adding them again.
+    if (isPointerEvent(event)) {
+      this._element.setPointerCapture(event.pointerId);
+      this._element.addEventListener('pointermove', this._move);
+      this._element.addEventListener('pointerup', this._pointerEnd);
+    } else { // MouseEvent
+      window.addEventListener('mousemove', this._move);
+      window.addEventListener('mouseup', this._pointerEnd);
+    }
+  }
+
+  /**
+   * Listener for touchstart. Bound to the class in the constructor.
+   * Only used if the browser doesn't support pointer events.
+   */
+  private _touchStart (event: TouchEvent) {
+    for (const touch of Array.from(event.changedTouches)) {
+      this._triggerPointerStart(new Pointer(touch), event);
+    }
+  }
+
+  /**
+   * Listener for pointer/mouse/touch move events.
+   * Bound to the class in the constructor.
+   */
+  private _move (event: PointerEvent | MouseEvent | TouchEvent) {
+    const previousPointers = this.currentPointers.slice();
+    const changedPointers = ('changedTouches' in event) ?
```

Comment:
Can you use `isPointerEvent` here?

Reply:
This is checking for a touch event, not a pointer event.

I could replace this with: `self.TouchEvent && event instanceof TouchEvent`, but I went for the property check since it's smaller.